src/VectorStore.ts

```diff
@@ -1,186 +1,292 @@
 
-	getByFilePath(filePath: string): Vector | undefined {
-		return this.filePathToVector.get(filePath)
+	async addFile(file: TFile) {
+		const existingFile = this.embeddings.get(file.path)
+		if (existingFile) return
+		const fileContents = await this.vault.read(file)
+		const hash = this.generateMd5Hash(fileContents)
+		const fileEntry = await this.generateFileEntry(fileContents, file.stat.mtime, hash)
+		if (fileEntry) {
+			this.embeddings.set(file.path, fileEntry)
+		}
+		await this.saveEmbeddingsToDatabaseFile() // todo debounce
 	}
 
-	async addVector(filePath: string, vector: Vector) {
-		this.filePathToVector.set(filePath, vector)
-		await this.saveVectorFile()
+	async updateFile(file: TFile) {
+		const existingFile = this.embeddings.get(file.path)
+		if (existingFile) {
+			const fileContents = await this.vault.read(file)
+			const hash = this.generateMd5Hash(fileContents)
+			if (hash !== existingFile.md5hash) {
+				const fileEntry = await this.generateFileEntry(fileContents, file.stat.mtime, hash)
+				if (fileEntry) {
+					this.embeddings.set(file.path, fileEntry)
+				}
+			} else {
+				console.log('no change, not updating db')
+			}
+		} else {
+			await this.addFile(file)
+		}
+		await this.saveEmbeddingsToDatabaseFile() // todo debounce
 	}
 
-	async updateVectorByFilename(filePath: string, updatedVector: Vector) {
-		this.filePathToVector.set(filePath, updatedVector)
-		await this.saveVectorFile()
+	async deleteFileByPath(filePath: string) {
+		this.embeddings.delete(filePath)
+		await this.saveEmbeddingsToDatabaseFile() // todo debounce
 	}
 
-	async updateFilename(oldFilePath: string, newFilePath: string, newVector: Vector) {
-		this.filePathToVector.delete(oldFilePath)
-		this.filePathToVector.set(newFilePath, newVector)
-		await this.saveVectorFile()
+	private generateMd5Hash(content: string) {
+		return Md5.hashStr(content)
 	}
 
-	async deleteByFilePath(filepath: string) {
-		const vector = this.filePathToVector.get(filepath)
-		if(vector != undefined) {
-			this.filePathToVector.delete(filepath)
-			await this.saveVectorFile()
+	private async chunkFile(fileContents: string): Promise<string[] | undefined> {
+		const messageRequest: ChatCompletionRequestMessage = {
+			role: ChatCompletionRequestMessageRoleEnum.User,
+			content: `I am indexing a file for search. 
+			When I search for a term, I want to be able to find the most relevant chunk of content from the file. 
+			To do that, I need to first break the file into topical chunks so that I can create embeddings for each chunk. 
+			When I search, I will compute the cosine similarity between the chunks and my search term and return chunks that are nearest. 
+			Please break the following file into chunks that would suit my use case. 
+			When you tell me the chunks you have decided on, include the original content from the file. Do not summarize. 
+			Prefix each chunk with "<<<CHUNK START>>>" so that I know where they begin. Here is the file:
+			${fileContents}`
```

Comment:
So far this prompt has always worked, but ChatGPT responses are inconsistent. What if the format returned is wrong? Could we even detect it? Maybe there is no way to know. One example is big files that don't ever say <<<CHUNK START>>> in them. Also, if the text returned isn't at least the same size as the original file then it didn't work! That seems like a good basic safeguard. The biggest potential risk I've seen is ChatGPT summarizing the original text in its response. We could also check if the chunk text is in the original file text. 🤔